src/github_sdk.py

```diff
@@ -1,101 +1,135 @@
-    envelope.add_transaction(trace)
-    base_uri, project_id = SENTRY_GITHUB_DSN.rsplit("/", 1)
-    sentry_key = base_uri.rsplit("@")[0].rsplit("https://")[1]
-    headers = {
-        "event_id": get_uuid(),  # Does this have to match anything?
-        "sent_at": format_timestamp(datetime.utcnow()),
-        "Content-Type": "application/x-sentry-envelope",
-        "Content-Encoding": "gzip",
-        "X-Sentry-Auth": f"Sentry sentry_key={sentry_key},"
-        + f"sentry_client=gha-sentry/0.0.1,sentry_timestamp={str(datetime.utcnow())},"
-        + "sentry_version=7",
-    }
-
-    # '{BASE_URI}/api/{PROJECT_ID}/{ENDPOINT}/'
-    url = f"{base_uri}/api/{project_id}/envelope/"
-    if GH_TOKEN and url.find("github.com") >= 0:
-        headers["Authorization"] = f"token {GH_TOKEN}"
-
-    body = io.BytesIO()
-    with gzip.GzipFile(fileobj=body, mode="w") as f:
-        envelope.serialize_into(f)
-
-    req = requests.post(url, data=body.getvalue(), headers=headers)
-    req.raise_for_status()
-
-
-# XXX: This is a slow call
-def get_extra_metadata(job):
-    runs = get(job["run_url"]).json()
-    workflow = get(runs["workflow_url"]).json()
-    repo = runs["repository"]["full_name"]
-    meta = {
-        # "workflow_name": workflow["name"],
-        "author": runs["head_commit"]["author"],
-        # https://getsentry.atlassian.net/browse/TET-22
-        # Tags are not linkified externally, plain text data can be selected in browsers and opened
-        "data": {
-            "job": job["html_url"],
-        },
-        "tags": {
-            "job_status": job["conclusion"],  # e.g. success, failure, skipped
-            "branch": runs["head_branch"],
-            "commit": runs["head_sha"],
-            "repo": repo,
-            "run_attempt": runs["run_attempt"],  # Rerunning a job
-            # It allows querying jobs within the same workflow (e.g. foo.yml)
-            "workflow": workflow["path"].rsplit("/")[-1],
-        },
-    }
-    if runs.get("pull_requests"):
-        pr_number = runs["pull_requests"][0]["number"]
-        meta["data"]["pr"] = f"https://github.com/{repo}/pull/{pr_number}"
-        meta["tags"]["pull_request"] = pr_number
-
-    return meta
+class GithubClient:
+    # This transform GH jobs conclusion keywords to Sentry performance status
+    github_status_trace_status = {"success": "ok", "failure": "internal_error"}
+
+    def __init__(self, dsn=None, github_token=None, dry_run=False) -> None:
+        self.token = github_token
+
+        # XXX: Make dsn mandatory in the future
+        if not dsn and not dry_run:
+            raise GithubSentryError("Set SENTRY_GITHUB_SDN in order to send envelopes.")
+
+        if dsn:
+            base_uri, project_id = dsn.rsplit("/", 1)
+            self.sentry_key = base_uri.rsplit("@")[0].rsplit("https://")[1]
+            # '{BASE_URI}/api/{PROJECT_ID}/{ENDPOINT}/'
+            self.sentry_project_url = f"{base_uri}/api/{project_id}/envelope/"
```

Comment:
Move this block out of the `send_envelope` function since we only need to execute it once.